# Import Pandas

In [3]:
# Import Pandas library for data manipulation and analysis

import pandas as pd

Data Exploration and Cleaning Using Pandas

Step 1: Loading the Dataset

In [4]:
#First I imported Pandas and loaded the CSV file into a DataFrame.

df = pd.read_csv("Combined_dataset.csv")

Step 2: Exploring the Dataset

In [6]:
# View first 5 rows of the dataset

df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


In [7]:
# View last 5 rows of the dataset

df.tail()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
995,13698258,KREDO,Men FORTE Silver Automatic Skeleton Analogue W...,3.5,11,11890,50.0,"""₹11,890.00""",INR,http://assets.myntassets.com/assets/images/136...,...,"{""1_star"":3,""2_stars"":0,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""pi-13698258""]",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
996,16896006,GUESS,Women White Patterned Dial & White Straps Anal...,5.0,22,8900,45.0,"""₹8,900.00""",INR,http://assets.myntassets.com/assets/images/168...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-16896006_cae700"",""https://videos.myntasse...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,...,"{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,...,"{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches
999,18601682,BOLDFIT,Purple Textured Yoga Mat,3.8,161,699,31.0,"""₹699.00""",INR,http://assets.myntassets.com/assets/images/186...,...,"{""1_star"":24,""2_stars"":8,""3_stars"":22,""4_stars...",NaN,NaN,"[{""size"":""Medium""}]",NaN,NaN,"[{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",yoga-mats


In [8]:
# Number of rows and columns

df.shape

(1000, 24)

In [9]:
# Display column names

df.columns

Index(['product_id', 'title', 'product_description', 'rating', 'ratings_count',
       'initial_price', 'discount', 'final_price', 'currency', 'images',
       'delivery_options', 'product_details', 'breadcrumbs',
       'product_specifications', 'amount_of_stars', 'what_customers_said',
       'seller_name', 'sizes', 'videos', 'seller_information', 'variations',
       'best_offer', 'more_offers', 'category'],
      dtype='str')

In [10]:
# Data types and overall information

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   str    
 2   product_description     1000 non-null   str    
 3   rating                  1000 non-null   float64
 4   ratings_count           1000 non-null   int64  
 5   initial_price           1000 non-null   int64  
 6   discount                879 non-null    float64
 7   final_price             1000 non-null   str    
 8   currency                1000 non-null   str    
 9   images                  1000 non-null   str    
 10  delivery_options        1000 non-null   str    
 11  product_details         1000 non-null   str    
 12  breadcrumbs             1000 non-null   str    
 13  product_specifications  1000 non-null   str    
 14  amount_of_stars         1000 non-null   str    
 15 

In [11]:
# Display data type of every column

df.dtypes

product_id                  int64
title                         str
product_description           str
rating                    float64
ratings_count               int64
initial_price               int64
discount                  float64
final_price                   str
currency                      str
images                        str
delivery_options              str
product_details               str
breadcrumbs                   str
product_specifications        str
amount_of_stars               str
what_customers_said           str
seller_name                   str
sizes                         str
videos                        str
seller_information            str
variations                    str
best_offer                    str
more_offers                   str
category                      str
dtype: object

Step 3: Handling Missing Values

In [12]:
#To identify missing values in the dataset:

df.isnull().sum()

product_id                  0
title                       0
product_description         0
rating                      0
ratings_count               0
initial_price               0
discount                  121
final_price                 0
currency                    0
images                      0
delivery_options            0
product_details             0
breadcrumbs                 0
product_specifications      0
amount_of_stars             0
what_customers_said       573
seller_name               301
sizes                       0
videos                    781
seller_information        301
variations                562
best_offer                  0
more_offers                 0
category                    0
dtype: int64

# After checking for missing values, I found that several columns contain null values.

- discount has 121 missing values.
- what_customers_said has 573 missing values.
- seller_name has 301 missing values.
- videos has 781 missing values.
- seller_information has 301 missing values.
- variations has 562 missing values.

Among these, videos and what_customers_said contain a large number of missing values.

Since columns such as videos and variations contain a large number of missing values and are not essential for this analysis, they were left unchanged.

For useful columns like discount, seller_name, and seller_information, missing values were filled appropriately.

In [13]:
# Fill missing discount values with the average discount

df["discount"] = df["discount"].fillna(df["discount"].mean())

# Fill missing seller names

df["seller_name"] = df["seller_name"].fillna("Unknown")

# Fill missing seller information

df["seller_information"] = df["seller_information"].fillna("Not Available")

In [15]:
#verify whether the above modified columns have modified or not

df[["discount","seller_name","seller_information"]].isnull().sum()

discount              0
seller_name           0
seller_information    0
dtype: int64

Step 4: Basic Operations

In [16]:
#Selecting specific columns

df[["title", "rating", "initial_price"]]

,title,rating,initial_price
0,Lino Perros,3.8,3995
1,Tommy Hilfiger,4.5,2899
2,Lavie,4.4,2999
3,F Gear,4.4,1675
4,MYTRIDENT,4.7,2899
...,...,...,...
995,KREDO,3.5,11890
996,GUESS,5.0,8900
997,GIORDANO,2.8,7950
998,Fastrack,4.3,3295


In [17]:
#Finding highly rated products

max_rated = df[df["rating"] > 4]

max_rated.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.000000,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,Unknown,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",Not Available,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.000000,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,Unknown,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.000000,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,Unknown,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.000000,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,Unknown,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,Not Available,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe
5,17198778,H&M,Beige & Tan Solid Stoneware Soap Dispenser,4.5,42,1399,53.503982,"""₹1,399.00""",INR,http://assets.myntassets.com/assets/images/171...,...,"{""1_star"":2,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,H &███Hen███ & █████████eta█████████████████████,"[{""size"":""Onesize""}]",NaN,Pra███yat███fra█████████PVT███████████████████...,"[{},{}]","{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",bathroom-accessories


In [18]:
#Finding products with large discounts

max_discount = df[df["discount"] > 50]

max_discount.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.000000,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,Unknown,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",Not Available,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.000000,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,Unknown,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.000000,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,Unknown,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
5,17198778,H&M,Beige & Tan Solid Stoneware Soap Dispenser,4.5,42,1399,53.503982,"""₹1,399.00""",INR,http://assets.myntassets.com/assets/images/171...,...,"{""1_star"":2,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,H &███Hen███ & █████████eta█████████████████████,"[{""size"":""Onesize""}]",NaN,Pra███yat███fra█████████PVT███████████████████...,"[{},{}]","{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",bathroom-accessories
6,19851824,AVI Living,Grey Solid 500 GSM Pure Cotton Bath Towel,4.0,5,1399,65.000000,"""₹1,399.00""",INR,http://assets.myntassets.com/assets/images/198...,...,"{""1_star"":0,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,Unknown,"[{""size"":""Onesize""}]",NaN,Not Available,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",bath-towels


# Step 5: Removing Duplicate Records


To ensure data quality, I checked whether the dataset contained any duplicate records. Duplicate rows can lead to incorrect analysis and should be removed before further processing.

In [22]:
#First, I checked whether duplicate records existed.

print(df.duplicated().sum())

0


# Observation:

The output shows that there are no duplicate records present in the dataset.

In [20]:
#Then I removed them.

df = df.drop_duplicates()

In [23]:
#Verification:

print(df.duplicated().sum())

0


Verification:

After applying the duplicate removal operation, the dataset still contains 0 duplicate records, confirming that no duplicate rows were present initially.

# Step 6: Creating a Derived Column

The assignment requires creating a new column called **total_amount** using the formula:

total_amount = price × quantity

Before performing this calculation, the **final_price** column needs to be cleaned because the values are stored as text and contain special characters such as quotation marks and commas. These values must be converted into a numeric format so that mathematical operations can be performed.

In [25]:
# Clean the final_price column

df["final_price"] = (
    df["final_price"]
    .astype(str)
    .str.replace("₹", "", regex=False)
    .str.replace(",", "", regex=False)
    .str.replace('"', "", regex=False)
    .str.strip()
)

# Convert to numeric datatype

df["final_price"] = pd.to_numeric(df["final_price"], errors="coerce")

# Creating Quantity and Total Amount Columns

The dataset does not contain a quantity column. For demonstration purposes, a quantity value of 1 is assigned to each product. Using this quantity value and the cleaned price column, a new column called total_amount is created.

In [26]:
# Create quantity column

df["quantity"] = 1

# Create total_amount column

df["total_amount"] = df["final_price"] * df["quantity"]

In [27]:
# Display the newly created columns

df[["final_price", "quantity", "total_amount"]].head()

,final_price,quantity,total_amount
0,3995.0,1,3995.0
1,2899.0,1,2899.0
2,2999.0,1,2999.0
3,1675.0,1,1675.0
4,2899.0,1,2899.0


# Observation:

The final_price column was successfully converted into a numeric format. A new column named total_amount was created using the formula:

total_amount = final_price × quantity

This derived column can be used for further analysis and calculations.

# Step 7: Saving the Cleaned Dataset

After completing the cleaning process, I saved the updated dataset as a new CSV file.

In [29]:
df.to_csv("cleaned_data.csv", index=False)

In [30]:
print("Dataset saved successfully")

Dataset saved successfully


# Brief Summary

1. Loaded the combined dataset into Pandas.
2. Explored the dataset using head(), tail(), shape(), columns, and info().
3. Identified and handled missing values.
4. Performed filtering and column selection operations.
5. Checked and removed duplicate records.
6. Created a new column called total_amount.
7. Saved the cleaned dataset as cleaned_dataset.csv.

Additional Findings:
- Several columns contained missing values.
- Price values required cleaning before numerical operations.
- No duplicate records were found.